src/zenml/lineage_graph/lineage_graph.py

```diff
@@ -39,17 +45,124 @@ class LineageGraph(BaseModel):
     root_step_id: Optional[str] = None
     run_metadata: List[Tuple[str, str, str]] = []
 
+    def generate_run_nodes_and_edges(
+        self, run: "PipelineRunResponseModel"
+    ) -> None:
+        """Initializes a lineage graph from a pipeline run.
+
+        Args:
+            run: The PipelineRunResponseModel to generate the lineage graph for.
+        """
+        self.run_metadata = [
+            (m.key, str(m.value), str(m.type)) for m in run.metadata.values()
+        ]
+
+        for step in run.steps.values():
+            self.generate_step_nodes_and_edges(step)
+
+        self.add_external_artifacts(run)
+        self.add_direct_edges(run)
+
     def generate_step_nodes_and_edges(
-        self, step: StepRunResponseModel
+        self, step: "StepRunResponseModel"
     ) -> None:
-        """Generates the step nodes and the edges between them.
+        """Generates the nodes and edges for a step and its artifacts.
 
         Args:
             step: The step to generate the nodes and edges for.
         """
         step_id = STEP_PREFIX + str(step.id)
+
+        # Set a root step if it doesn't exist yet
         if self.root_step_id is None:
             self.root_step_id = step_id
+
+        # Add the step node
+        self.add_step_node(step, step_id)
+
+        # Add nodes and edges for all output artifacts
+        for artifact_name, artifact in step.outputs.items():
+            artifact_id = ARTIFACT_PREFIX + str(artifact.id)
+            self.add_artifact_node(
+                artifact=artifact,
+                id=artifact_id,
+                name=artifact_name,
+                step_id=str(step_id),
+                status=step.status,
+            )
+            self.add_edge(step_id, artifact_id)
+
+        # Add nodes and edges for all input artifacts
+        for artifact_name, artifact in step.inputs.items():
+            artifact_id = ARTIFACT_PREFIX + str(artifact.id)
+            self.add_edge(artifact_id, step_id)
+
+    def add_external_artifacts(self, run: "PipelineRunResponseModel") -> None:
+        """Adds all external artifacts to the lineage graph.
+
+        Args:
+            run: The pipeline run to add external artifacts for.
+        """
+        nodes_ids = {node.id for node in self.nodes}
+        for step in run.steps.values():
+            for artifact_name, artifact in step.inputs.items():
+                artifact_id = ARTIFACT_PREFIX + str(artifact.id)
+                if artifact_id not in nodes_ids:
+                    self.add_artifact_node(
+                        artifact=artifact,
+                        id=artifact_id,
+                        name=artifact_name,
+                        step_id=str(artifact.producer_step_run_id),
+                        status="External",
```

Comment:
Feels a bit like we are misusing the status field here in order to make artifacts fit the mold. Especially with the loosening of the BaseNodeDetails.status typing. Ideally I would keept this as an enum at least so we have a source of truth of allowed statuses.

Reply:
Long-term the entire DAG creation should probably move to the frontend, IMO it doesn't make sense that the backend is doing this behind a separate endpoint since all the required information is already contained in the `PipelineRunResponseModel`.